File: app/js/seni/RenderPacket.js
/*
 *  Seni
 *  Copyright (C) 2016 Inderjit Gill <email@indy.io>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

// the size of transferrable data held by each RenderPacket
const renderPacketSize = 1 * 1024 * 1024; // 1MB

// number of float32 elements in each data type
const vertexItemSize = 2; // xy
const colourItemSize = 4; // rgba

const f32Size = 4; // 4 bytes in a float32

// each buffer can hold bufferSize 'items' where each item is a vertex+colour
const itemSize = f32Size * (vertexItemSize + colourItemSize);
const bufferSize = parseInt(renderPacketSize / itemSize, 10);

export default class RenderPacket {
  constructor() {
    // pass in sizes so that GLExec can access them
    this.vertexItemSize = vertexItemSize;
    this.colourItemSize = colourItemSize;

    this.abVertex = new ArrayBuffer(vertexItemSize * f32Size * bufferSize);
    this.abColour = new ArrayBuffer(colourItemSize * f32Size * bufferSize);

    this.vertexBuffer = new Float32Array(this.abVertex);
    this.colourBuffer = new Float32Array(this.abColour);
    // the level of both the vertex and colour buffer
    // to find the actual index position multiply bufferLevel
    // by the relevant itemSize of the buffer
    this.bufferLevel = 0;
    this.vertexBufferLevel = 0;
    this.colourBufferLevel = 0;
  }

  // can the number of vertices fit into this RenderPacket?
  canVerticesFit(numVertices) {
    return this.bufferLevel < bufferSize - (numVertices + 2);
  }

  getBufferLevel() {
    return this.bufferLevel;
  }

  isRenderPacketEmpty() {
    return this.bufferLevel === 0;
  }

  formDegenerateTriangle(p0) {
    // add two vertex entries which will form degenerate triangles

    // get the index of the last vertex that was added
    const index = this.vertexBufferLevel - vertexItemSize;
    // just copy the previous entries
    // note: colour doesn't matter since these triangles won't be rendered
    this.addVertex([this.vertexBuffer[index + 0], this.vertexBuffer[index + 1]],
                   [0, 0, 0, 0]);

    // add the new vertex to complete the degenerate triangle
    this.addVertex(p0, [0, 0, 0, 0]);

    // Note: still need to call addVertex on the first
    // vertex when we 'really' render the strip
  }

  // appends the vertex with position p and colour c
  addVertex(p, c) {
    this.vertexBuffer[this.vertexBufferLevel + 0] = p[0];
    this.vertexBuffer[this.vertexBufferLevel + 1] = p[1];

    this.colourBuffer[this.colourBufferLevel + 0] = c[0];
    this.colourBuffer[this.colourBufferLevel + 1] = c[1];
    this.colourBuffer[this.colourBufferLevel + 2] = c[2];
    this.colourBuffer[this.colourBufferLevel + 3] = c[3];

    this.bufferLevel += 1;
    this.vertexBufferLevel += vertexItemSize;
    this.colourBufferLevel += colourItemSize;
  }
}

File: core/src/render_packet.rs
// Copyright (C) 2020 Inderjit Gill <email@indy.io>

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use crate::error::{Error, Result};
use crate::matrix::Matrix;
use crate::rgb::Rgb;

pub const RENDER_PACKET_MAX_SIZE: usize = 262_144;
pub const RENDER_PACKET_FLOAT_PER_VERTEX: usize = 8;
// 262144 * 4 == 1MB per render packet
// 262144 / 8 == 32768 vertices per render packet

#[derive(Default)]
pub struct RenderPacketGeometry {
    pub geo: Vec<f32>,
}

#[derive(Default)]
pub struct RenderPacketMask {
    pub filename: String,
    pub invert: bool,
}

// the final image
pub struct RenderPacketImage {
    pub linear_colour_space: bool,
    pub contrast: f32,
    pub brightness: f32,
    pub saturation: f32,
}

pub enum RenderPacket {
    Geometry(RenderPacketGeometry),
    Mask(RenderPacketMask),
    Image(RenderPacketImage),
}

impl RenderPacket {
    pub fn get_mut_render_packet_geometry(&mut self) -> Result<&mut RenderPacketGeometry> {
        match self {
            RenderPacket::Geometry(rpg) => Ok(rpg),
            _ => Err(Error::Geometry),
        }
    }
}

impl RenderPacketMask {
    pub fn new() -> RenderPacketMask {
        RenderPacketMask {
            filename: "".to_string(),
            invert: false,
        }
    }
}

impl RenderPacketGeometry {
    pub fn new() -> RenderPacketGeometry {
        RenderPacketGeometry {
            geo: Vec::with_capacity(RENDER_PACKET_MAX_SIZE),
        }
    }

    pub fn get_geo_len(&self) -> usize {
        self.geo.len()
    }

    pub fn get_geo_ptr(&self) -> *const f32 {
        self.geo.as_ptr() as *const f32
    }

    pub fn add_vertex(&mut self, matrix: &Matrix, x: f32, y: f32, col: &Rgb, u: f32, v: f32) {
        // assuming that col is ColourFormat::Rgb

        let (nx, ny) = matrix.transform_vec2(x, y);

        // note: the shader should pre-multiply the r,g,b elements by alpha
        self.geo
            .append(&mut vec![nx, ny, col.0, col.1, col.2, col.3, u, v]);
    }

    pub fn form_degenerate_triangle(&mut self, matrix: &Matrix, x: f32, y: f32) {
        // just copy the previous entries
        self.dup();

        // add the new vertex to complete the degenerate triangle
        let rgb = Rgb::new(0.0, 0.0, 0.0, 0.0);
        self.add_vertex(matrix, x, y, &rgb, 0.0, 0.0);

        // Note: still need to call addVertex on the first
        // vertex when we 'really' render the strip
    }

    // duplicate the last geometry point
    pub fn dup(&mut self) {
        let len = self.geo.len();

        let x: f32;
        let y: f32;
        let r: f32;
        let g: f32;
        let b: f32;
        let a: f32;
        let u: f32;
        let v: f32;
        {
            x = self.geo[len - 8];
            y = self.geo[len - 7];
            r = self.geo[len - 6];
            g = self.geo[len - 5];
            b = self.geo[len - 4];
            a = self.geo[len - 3];
            u = self.geo[len - 2];
            v = self.geo[len - 1];
        }

        self.geo.append(&mut vec![x, y, r, g, b, a, u, v]);
    }

    pub fn can_vertices_fit(&self, num: usize) -> bool {
        self.geo.len() < (RENDER_PACKET_MAX_SIZE - (num * RENDER_PACKET_FLOAT_PER_VERTEX))
    }

    pub fn is_empty(&self) -> bool {
        self.geo.is_empty()
    }
}
